### PR TITLE
fixes #3284 updates tablet locs using conditional mutations

### LINKED ISSUE #3284: Set tablet locations using conditional mutations
After #3251 is merged into the elasticity branch changes can be made to set the tablet location using conditional mutations.  These changes could also be made in the elasticity branch.  Switching accumulo metadata operations to use conditional mutations will have to be done piecemeal and setting the locations is a good place to start.

The following work is what would likely need to be done.

 * In the mananger set the future location with a condition that no location is set for the tablet AND no operation is active on the tablet.
 * In the tablet server set the current location with a condition that the future location is the tserver AND no operation is active on the tablet.
 * In the manager for the case of a dead tserver delelete the current location with a condition that the dead tserver is the current location AND no operation is active on the tablet.

### PR DESCRIPTION
**Pull request by keith-turner**
Updates tablet locations using conditional mutations. Two supporting changes were made and one bug was fixes while making this change.

The first supporting change was streamlining handling of conditional mutations with a result of UNKNOWN.  An UNKNOWN result on a conditional mutation occurs when the RPC for the conditional mutation has an error.  In this case the conditional mutation may or may not have gone through. The tablet must be read to know what happened.  This update adds support for automatically reading the tablet and checking it via a lambda. This makes it easy to write code for handling the unknown case.

The second supporting change was combining code that was mostly the same in ZooTabletStateStore and MetaDataStateStore by making both extend AbstractStateStore and use common code.  This change allowed the updates to use conditional mutations to be made in one place instead of two.

The bug was with the new conditional writer code, it only supported writing tablets of the same table.  The code was changed to only require that tablets be on the same datalevel.  This change allowed ITs that create multiple tables to run.

Some places in the code that set locations were not changed to use conditional mutations.  Comments were placed in the code for these.

### CODE REVIEW

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/Ample.java [hunk 1]
```diff
@@ -398,9 +399,7 @@ ConditionalTabletMutator requireOperation(TabletOperation operation,
   /**
    * Convenience interface for handling conditional mutations with a status of UNKNOWN.
    */
-  interface UknownValidator {
-    boolean shouldAccept(TabletMetadata tabletMetadata);
-  }
+  interface UknownValidator extends Predicate<TabletMetadata> {}
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/metadata/schema/Ample.java:402]: Should this typo be fixed?

```suggestion
  interface UnknownValidator extends Predicate<TabletMetadata> {}
```

File: server/base/src/main/java/org/apache/accumulo/server/manager/state/AbstractTabletStateStore.java [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+
+  @Override
+  public void setLocations(Collection<Assignment> assignments) throws DistributedStoreException {
+    try (var tabletsMutator = ample.conditionallyMutateTablets()) {
+      for (Assignment assignment : assignments) {
+        var conditionalMutator = tabletsMutator.mutateTablet(assignment.tablet)
+            .requireAbsentOperation()
+            .requireLocation(TabletMetadata.Location.future(assignment.server))
+            .putLocation(TabletMetadata.Location.current(assignment.server))
+            .deleteLocation(TabletMetadata.Location.future(assignment.server)).deleteSuspension();
```
Comment (dlmarion) [inline server/base/src/main/java/org/apache/accumulo/server/manager/state/AbstractTabletStateStore.java:56]: It appears that `ConditionalWriter.convertConditions` will sort the Conditions based on CF, CQ, Visibility, and then timestamp. Given that `requireLocation`, `putLocation`, and `deleteLocation` all operate on the same CF and CQ with a timestamp of 0L, do we know in which order they are applied?

Comment (keith-turner) [reply server/base/src/main/java/org/apache/accumulo/server/manager/state/AbstractTabletStateStore.java:56]: The requireLocation call will create a condition on the conditional mutation.  All conditions on a conditional mutation are checked first before any changes are made.  So a check is done to see if the future location is as expected and if it is then the changes to delete the future and set the current are made.   If the check of the current future location fails then no changes are made.

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/Ample.java (hunk 1 above)
Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/metadata/schema/Ample.java]: updated in c5fe3e2

File: server/base/src/main/java/org/apache/accumulo/server/manager/state/AbstractTabletStateStore.java (hunk 2 above)
Comment (keith-turner) [reply server/base/src/main/java/org/apache/accumulo/server/manager/state/AbstractTabletStateStore.java:56]: So what happens on the server is

 1. it gets a lock on the row
 2. it checks the cf=future and cq=session match what was specified in requireLocation(TabletMetadata.Location.future(assignment.server))
 3. If the above check was succesful it deletes the future col and sets the current col
 4. releases the row lock

Comment (dlmarion) [reply server/base/src/main/java/org/apache/accumulo/server/manager/state/AbstractTabletStateStore.java:56]: Sorry for asking so many questions about this, I'm not familiar with ConditionalMutations. In the code above there is an implied order in the conditions, but it appears that they are re-sorted on the client side before sending to the server [here](https://github.com/apache/accumulo/blob/ba472d6e24daa8f0014a22cabace3061f5d46413/core/src/main/java/org/apache/accumulo/core/clientImpl/ConditionalWriterImpl.java#L777) which is called as part of the write process in `ConditionalWriterImpl.sendToServer`. 

Are all of the conditions checked before any changes are applied and therefore the order is irrelevant?

Comment (keith-turner) [reply server/base/src/main/java/org/apache/accumulo/server/manager/state/AbstractTabletStateStore.java:56]: > Are all of the conditions checked before any changes are applied and therefore the order is irrelevant?

That's correct.  The code that sorts the conditions just does that to make them more efficient to check.

Comment (dlmarion) [reply server/base/src/main/java/org/apache/accumulo/server/manager/state/AbstractTabletStateStore.java:56]: ok, thanks. I think that alleviates my concerns.

**Pull request merged**